domain/credential/state/credential.go

```diff
@@ -0,0 +1,489 @@
+type State struct {
+	*domain.StateBase
+}
+
+// NewState creates a state to access the database.
+func NewState(factory coredatabase.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// UpsertCloudCredential adds or updates a cloud credential with the given tag.
+func (st *State) UpsertCloudCredential(ctx context.Context, name, cloud, owner string, credential cloud.Credential) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		// Get the credential UUID - either existing or make a new one.
+		// TODO(wallyworld) - implement owner as a FK when users are modelled.
+		stmt := `
+SELECT  cloud_credential.uuid
+FROM    cloud_credential
+        INNER JOIN cloud
+            ON cloud_credential.cloud_uuid = cloud.uuid
+WHERE   cloud.name = ? AND cloud_credential.name = ? AND cloud_credential.owner_uuid = ?
+`
+		var credentialUUID string
+		row := tx.QueryRowContext(ctx, stmt, cloud, name, owner)
+		err := row.Scan(&credentialUUID)
+		if err != nil && err != sql.ErrNoRows {
+			return errors.Trace(err)
+		}
+		if err != nil {
+			if credential.Invalid || credential.InvalidReason != "" {
+				return fmt.Errorf("adding invalid credential %w", errors.NotSupported)
+			}
+			credentialUUID = utils.MustNewUUID().String()
+		}
+
+		if err := upsertCredential(ctx, tx, credentialUUID, name, cloud, owner, credential); err != nil {
+			return errors.Annotate(err, "updating credential")
+		}
+
+		if err := updateCredentialAttributes(ctx, tx, credentialUUID, credential.Attributes()); err != nil {
+			return errors.Annotate(err, "updating credential attributes")
+		}
+
+		// TODO(wallyworld) - update model status (suspended etc)
+
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func upsertCredential(ctx context.Context, tx *sql.Tx, credentialUUID string, name, cloud, owner string, credential cloud.Credential) error {
+	dbCredential, err := dbCredentialFromCredential(ctx, tx, credentialUUID, name, cloud, owner, credential)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+INSERT INTO cloud_credential (uuid, name, cloud_uuid, auth_type_id, owner_uuid, revoked, invalid, invalid_reason)
+VALUES (?, ?, ?, ?, ?, ?, ?, ?)
+ON CONFLICT(uuid) DO UPDATE SET name=excluded.name,
+                                cloud_uuid=excluded.cloud_uuid,
+                                auth_type_id=excluded.auth_type_id,
+                                owner_uuid=excluded.owner_uuid,
+                                revoked=excluded.revoked,
+                                invalid=excluded.invalid,
+                                invalid_reason=excluded.invalid_reason;`
+
+	_, err = tx.ExecContext(ctx, q,
+		dbCredential.ID,
+		dbCredential.Name,
+		dbCredential.CloudUUID,
+		dbCredential.AuthTypeID,
+		dbCredential.OwnerUUID,
+		dbCredential.Revoked,
+		dbCredential.Invalid,
+		dbCredential.InvalidReason,
+	)
```

Comment:
This would be much easier with sqlair.

Reply:
"Much easier" is debatable IMO. I've re-implemented with sqlair and it's almost as verbose. You do save a little on bind args.

Reply:
There's a pending PR which will improve things https://github.com/canonical/sqlair/pull/69